### PR TITLE
[Fixes #207] Share projects

### LINKED ISSUE #207: Share projects
## Feature Request

Is your feature request related to an existing [Issue](https://www.github.com/Auto-DL/Generator/issues) or a new implementation?

### Feature Description

Add "Share" option in the below image, which shares the to project another user and lets discuss the approach with Aditya , Tanvi and Raj (me)

image

### PR DESCRIPTION
**Pull request by therajtiwari**
# Pull Request

## What does this PR do?
Fixes #207

- Actions now have a share option. 
- A user can share a project with all available users and can view all of the users with whom he has shared it.
- Shared projects are shown separately from personal projects in a table below and can be edited and deleted.

## UI
Screenshot from 2021-07-13 12-25-48
Screenshot from 2021-07-13 12-27-05
Screenshot from 2021-07-13 12-28-10
Screenshot from 2021-07-13 12-29-03

## What part does this affect?
- [x] FrontEnd.
- [x] BackEnd.
- [ ] Documentation.
- [ ] Other. (Please specify below)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue or slack?
- [x] Did you read the [contributor guideline](https://github.com/Auto-DL/Auto-DL/blob/v1-beta/CONTRIBUTING.md)?
- [x] Did you ensure that there aren't any other open [Pull Requests](https://www.github.com/Auto-DL/Generator/pulls) for the same update/change?
- [x] Did you make sure the title is self-explanatory and the description concisely explains the PR?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure the code is clean and docstrings have been added or updated as required?
- [x] Did you make sure the code is linted/formatted locally prior to submission? (using `black` and/or `prettier`)
- [ ] Did you make sure to update the documentation with your changes? (if necessary)

## PR review
Anyone in the community is free to review the PR once the tests have passed.

Thank you for contributing to [AutoDL](https://github.com/Auto-DL/Auto-DL). We look forward to your continued support.

### CODE REVIEW

File: BackEndApp/v1/views.py [hunk 1]
```diff
@@ -604,3 +632,66 @@ def get_users(request):
...
+            dst = os.path.join(
+                store_obj.rootpath,
+                share_with,
+                "shared_" + str(project_id),
+            )
+
+            try:
+                os.symlink(src, dst)
+                if not metadata["shared_by"]:
+                    metadata["shared_by"] = shared_by
```
Comment (ADI10HERO) [inline BackEndApp/v1/views.py:680]:
```suggestion
                if not metadata.get("shared_by", None):
                    metadata["shared_by"] = shared_by
```

**Review by ADI10HERO — CHANGES_REQUESTED** [review]
Let's schedule a meet to discuss this PR

**Comment by therajtiwari** [comment]
> Let's schedule a meet to discuss this PR

Sure, let me know the timing.

File: BackEndApp/authv1/store.py [hunk 2]
```diff
@@ -32,7 +32,11 @@ def delete(self, project=None):
             _path = self.path
             if project is not None:
                 _path = posixpath.join(_path, project)
-            shutil.rmtree(_path)
+            if project.startswith("shared_"):
+                os.remove(_path)
+            else:
+                shutil.rmtree(_path)
```
Comment (ADI10HERO) [inline BackEndApp/authv1/store.py:38]:
```suggestion
            os.remove(_path)
```

File: BackEndApp/v1/urls.py [hunk 3]
```diff
@@ -33,5 +33,7 @@
     # post-processing routes
     path("code/download/", views.download_code, name="dowload-code"),
     # Get users
-    path("users/get", views.get_users, name="get-users"),
+    # path("users/get", views.get_users, name="get-users"),
+    path("users/getall", views.get_all_users, name="get-all-users"),
```
Comment (ADI10HERO) [inline BackEndApp/v1/urls.py:37]:
```suggestion
    path("users/get/all/", views.get_all_users, name="get-all-users"),
```

**Review by ADI10HERO — APPROVED** [review]
Hmm looks good.

Just fix the merge conflict.

@rajraj889 please review the react part

File: BackEndApp/authv1/store.py (hunk 2 above)
Comment (ADI10HERO) [reply BackEndApp/authv1/store.py:38]: Oww, you can look at this suggestion or let me know if it won't work @therajtiwari

File: BackEndApp/v1/views.py (hunk 1 above)
Comment (ADI10HERO) [reply BackEndApp/v1/views.py]: Won't work without this @therajtiwari

**Review by ADI10HERO — CHANGES_REQUESTED** [review]
PLease have a look the suggested comments

**Review by rajraj889 — CHANGES_REQUESTED** [review]
Share modal doesn't close when a project is shared or when sharing is failed.
image

File: BackEndApp/authv1/store.py (hunk 2 above)
Comment (therajtiwari) [reply BackEndApp/authv1/store.py:38]: It actually won't work. Can you please check my Slack message?

**Comment by therajtiwari** [comment]
> Share modal doesn't close when a project is shared or when sharing is failed.
> image

Actually, I deliberately left it open as it would be convenient for the user to try again in case of a failure and to share with additional users after a successful share. How do you feel about this?

**Review by rajraj889 — APPROVED** [review]
Okay, let's try this and observe users reactions.

**Review by ADI10HERO — APPROVED** [review]
Looks good...
Let's solve the bugs as we find them :P

**Pull request merged**